### PR TITLE
Add discouraged_init opt-in rule that discourages direct initialization of certain types

### PR DESCRIPTION
**Pull request by ornithocoder**
Implements #1306, requested by  @Noobish1.

* Opt-in rule,
* Implements UIDevice() and Bundle() by default,
* Allows severity configuration (warning or error) via .swiftlint,
* Allows overwriting types via .swiftlint.

### CODE REVIEW

File: Source/SwiftLintFramework/Rules/DiscouragedInitRule.swift [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    )
+
+    public func validate(file: File,
+                         kind: SwiftExpressionKind,
+                         dictionary: [String: SourceKitRepresentable]) -> [StyleViolation] {
+        guard
+            let offset = dictionary.nameOffset,
+            let name = dictionary.name,
+            kind == .call,
+            dictionary.bodyLength == 0,
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/DiscouragedInitRule.swift]: so, this rule actually just works for inits without parameters, right?

In this case, I'd rename the rule and try to improve the description as well.

File: Source/SwiftLintFramework/Rules/DiscouragedInitRule.swift [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            configuration.discouragedInits.contains(name)
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/DiscouragedInitRule.swift]: what do you think about also checking for `Bundle.init()` for example?

File: Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedInitConfiguration.swift [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    public var consoleDescription: String {
+        return severityConfiguration.consoleDescription
+    }
+
+    public var severity: ViolationSeverity {
+        return severityConfiguration.severity
+    }
+
+    private(set) public var discouragedInits: [String]
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedInitConfiguration.swift]: what do you think about making it a `Set<String>`?

File: Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedInitConfiguration.swift [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+//
+
+import Foundation
+import SourceKittenFramework
+
+public struct DiscouragedInitConfiguration: RuleConfiguration, Equatable {
+    public var severityConfiguration = SeverityConfiguration(.warning)
+
+    public var consoleDescription: String {
+        return severityConfiguration.consoleDescription
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedInitConfiguration.swift]: You need to include all configuration in the description. 
For example:

```swift
return "severity: \(severityConfiguration.consoleDescription), included: \(discouragedInits)"
```

File: Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedInitConfiguration.swift [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        guard let configuration = configuration as? [String: Any] else {
+            throw ConfigurationError.unknownConfiguration
+        }
+
+        if let severityString = configuration["severity"] as? String {
+            try severityConfiguration.apply(configuration: severityString)
+        }
+
+        if let included = [String].array(of: configuration["included"]) {
+            discouragedInits += included
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedInitConfiguration.swift]: I think this should always override the default values. Doing that, we provide a way to opt-out from the default configuration if needed (for example, one might have a class `Bundle` in their project).

File: Tests/SwiftLintFrameworkTests/RulesTests.swift [hunk 6]
```diff
@@ -62,6 +62,26 @@ class RulesTests: XCTestCase {
         verifyRule(DiscardedNotificationCenterObserverRule.description)
     }
 
+    func testDiscouragedInit() {
```
Comment (marcelofabri) [inline Tests/SwiftLintFrameworkTests/RulesTests.swift]: can you add a new test class and split this in several tests, please?

File: CHANGELOG.md [hunk 7]
```diff
@@ -92,6 +92,11 @@
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#1714](https://github.com/realm/SwiftLint/issues/1714)
 
+* Add `discouraged_init` opt-in rule that discourages direct
+  initialization of certain classes.
```
Comment (marcelofabri) [inline CHANGELOG.md]: I'd use "types" instead of "classes" here.

File: Source/SwiftLintFramework/Rules/DiscouragedInitRule.swift (hunk 1 above)
Comment (ornithocoder) [inline Source/SwiftLintFramework/Rules/DiscouragedInitRule.swift]: Precisely, it's meant to direct initialization only. I'll update the rule name and description.

File: Source/SwiftLintFramework/Rules/DiscouragedInitRule.swift (hunk 2 above)
Comment (ornithocoder) [inline Source/SwiftLintFramework/Rules/DiscouragedInitRule.swift]: Sure, I'll do that.

File: Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedInitConfiguration.swift (hunk 3 above)
Comment (ornithocoder) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedInitConfiguration.swift]: Much better! Changing it.

File: Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedInitConfiguration.swift (hunk 4 above)
Comment (ornithocoder) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedInitConfiguration.swift]: Sorry about that. Updating it.

File: Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedInitConfiguration.swift (hunk 5 above)
Comment (ornithocoder) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedInitConfiguration.swift]: No problem. I'll update the code.

File: Tests/SwiftLintFrameworkTests/RulesTests.swift (hunk 6 above)
Comment (ornithocoder) [inline Tests/SwiftLintFrameworkTests/RulesTests.swift]: Sure, but I might have to disable the type_body_length rule for this file, tho.

File: CHANGELOG.md (hunk 7 above)
Comment (ornithocoder) [inline CHANGELOG.md]: Yep!

File: Tests/SwiftLintFrameworkTests/RulesTests.swift (hunk 6 above)
Comment (marcelofabri) [inline Tests/SwiftLintFrameworkTests/RulesTests.swift]: I meant making it a new file. Sorry, should have been more explicit about it.

Comment (ornithocoder) [inline Tests/SwiftLintFrameworkTests/RulesTests.swift]: Sure, no problem. Should I leave the "default" one in this one?

Comment (marcelofabri) [inline Tests/SwiftLintFrameworkTests/RulesTests.swift]: I think it makes sense to move everything to the new file. See `FileHeaderRuleTests.swift` for example.

Comment (ornithocoder) [inline Tests/SwiftLintFrameworkTests/RulesTests.swift]: I'll move them all. Easier to maintain ;-)

File: Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedDirectInitConfiguration.swift [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    public mutating func apply(configuration: Any) throws {
+        guard let configuration = configuration as? [String: Any] else {
+            throw ConfigurationError.unknownConfiguration
+        }
+
+        if let severityString = configuration["severity"] as? String {
+            try severityConfiguration.apply(configuration: severityString)
+        }
+
+        if let included = [String].array(of: configuration["included"]) {
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedDirectInitConfiguration.swift]: actually, on a second thought, what do you think about using `"types"` as the key? I'm afraid that we already overload the term `included` too much.

File: Tests/SwiftLintFrameworkTests/DiscouragedDirectInitRuleTests.swift [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+//
+
+import Foundation
+import SwiftLintFramework
+import XCTest
+
+class DiscouragedDirectInitRuleTests: XCTestCase {
+    let baseDescription = DiscouragedDirectInitRule.description
+
+    func testDiscouragedDirectInitRuleWithDefaultConfiguration() {
```
Comment (marcelofabri) [inline Tests/SwiftLintFrameworkTests/DiscouragedDirectInitRuleTests.swift]: could you remove the `Rule` from the method names just to keep consistent with other tests?

**Comment by marcelofabri** [comment]
@ornithocoder This looks great! I just have two more small nitpicks 😅

File: Tests/SwiftLintFrameworkTests/DiscouragedDirectInitRuleTests.swift (hunk 9 above)
Comment (ornithocoder) [inline Tests/SwiftLintFrameworkTests/DiscouragedDirectInitRuleTests.swift]: Sure, removed.

File: Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedDirectInitConfiguration.swift (hunk 8 above)
Comment (ornithocoder) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedDirectInitConfiguration.swift]: It makes sense. Changed.

File: Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedDirectInitConfiguration.swift [hunk 10]
```diff
@@ -46,7 +46,7 @@ public struct DiscouragedDirectInitConfiguration: RuleConfiguration, Equatable {
             try severityConfiguration.apply(configuration: severityString)
         }
 
-        if let included = [String].array(of: configuration["included"]) {
+        if let included = [String].array(of: configuration["types"]) {
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedDirectInitConfiguration.swift]: You need to change on consoleDescription too

**Comment by jpsim** [comment]
Why is this opt-in?

File: Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedDirectInitConfiguration.swift (hunk 10 above)
Comment (ornithocoder) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/DiscouragedDirectInitConfiguration.swift]: Yep, and the variable name :-) Done.

**Comment by ornithocoder** [comment]
@jpsim @marcelofabri I didn't spend much time thinking of that and went for the easy option. Should I make it a default rule instead?

**Comment by marcelofabri** [comment]
I think that the chance of getting a false positive with `UIDevice` or `Bundle` is low, so 👍 on making this enabled by default.

**Comment by jpsim** [comment]
@marcelofabri you merged this but it's still marked as opt-in... 🤔 

did you mean to change that before merging?

**Comment by marcelofabri** [comment]
I'm opening a new PR to change it.

**Comment by marcelofabri** [comment]
Here it is: #1736. My plan was to merge it without a PR (🙈), but I've realized that actually `LinuxMain.swift` was outdated.

**Pull request merged**